modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java

```
@@ -0,0 +1,348 @@
+    public static int multiplyExact(int x, int y) {
+        long r = (long)x * (long)y;
+
+        return convertToIntExact(r);
+    }
+
+    /** Returns the product of the arguments, throwing an exception if the result overflows an {@code short}. */
+    public static short multiplyExact(short x, short y) {
+        int r = x * y;
+
+        if (r != (short)r)
+            throw new ArithmeticException(SMALLINT.getName() + " overflow");
+
+        return (short)r;
+    }
+
+    /** Returns the product of the arguments, throwing an exception if the result overflows an {@code byte}. */
+    public static byte multiplyExact(byte x, byte y) {
+        int r = x * y;
+
+        if (r != (byte)r)
+            throw new ArithmeticException(TINYINT.getName() + " overflow");
+
+        return (byte)r;
+    }
+
+    /** Returns the quotient of the arguments, throwing an exception if the result overflows an {@code long}. */
+    public static long divideExact(long x, long y) {
+        if (y == -1)
+            return negateExact(x);
+
+        return x / y;
+    }
+
+    /** Returns the quotient of the arguments, throwing an exception if the result overflows an {@code int}. */
+    public static int divideExact(int x, int y) {
+        if (y == -1)
+            return negateExact(x);
+
+        return x / y;
+    }
+
+    /** Returns the quotient of the arguments, throwing an exception if the result overflows an {@code short}. */
+    public static short divideExact(short x, short y) {
+        if (y == -1)
+            return negateExact(x);
+
+        return (short)(x / y);
+    }
+
+    /** Returns the quotient of the arguments, throwing an exception if the result overflows an {@code byte}. */
+    public static byte divideExact(byte x, byte y) {
+        if (y == -1)
+            return negateExact(x);
+
+        return (byte)(x / y);
+    }
+
+    /** */
+    private static void checkNumberLongBounds(SqlTypeName type, Number x) {
```

seems you can safely remove such a check or i miss smth ?

---

This check is required since BigDecimal.longValue, Double.longValue and Float.longValue don't check long value bounds and can produce wrong result. For example `new BigDecimal("18446744073709551616").longValue()` is 0, and this value passes all checks, but should throw an error.

---

ok be seems such cases are not covered by tests, or already done ?

---

It's covered by `CAST(9223372036854775807.5 + 9223372036854775807.5 AS INTEGER)` in test